packages/host/app/components/ai-assistant/formatted-message.gts

```diff
@@ -1,18 +1,98 @@
+import { service } from '@ember/service';
 import type { SafeString } from '@ember/template';
 import Component from '@glimmer/component';
 
+import { tracked } from '@glimmer/tracking';
+
+import { task } from 'ember-concurrency';
+import perform from 'ember-concurrency/helpers/perform';
+import { TrackedArray, TrackedObject } from 'tracked-built-ins';
+
+import ApplySearchReplaceBlockCommand from '@cardstack/host/commands/apply-search-replace-block';
 import CodeBlock from '@cardstack/host/modifiers/code-block';
 import { MonacoEditorOptions } from '@cardstack/host/modifiers/monaco';
 
+import type CardService from '@cardstack/host/services/card-service';
+import CommandService from '@cardstack/host/services/command-service';
+import LoaderService from '@cardstack/host/services/loader-service';
 import { type MonacoSDK } from '@cardstack/host/services/monaco-service';
 
+import ApplyButton from '../ai-assistant/apply-button';
+
 interface FormattedMessageSignature {
   sanitizedHtml: SafeString;
   monacoSDK: MonacoSDK;
   renderCodeBlocks: boolean;
 }
 
+interface CodeAction {
+  fileUrl: string;
+  code: string;
+  containerElement: HTMLDivElement;
+  state: 'ready' | 'applying' | 'applied' | 'failed';
+}
+
 export default class FormattedMessage extends Component<FormattedMessageSignature> {
+  @tracked codeActions: TrackedArray<CodeAction> = new TrackedArray([]);
+  @service private declare cardService: CardService;
+  @service private declare loaderService: LoaderService;
+  @service private declare commandService: CommandService;
+
+  registerCodeBlockContainer = (
+    fileUrl: string,
+    code: string,
+    codeActionContainerElement: HTMLDivElement,
+  ) => {
+    let searchReplaceRegex =
+      /<<<<<<< SEARCH\n(.*)\n=======\n(.*)\n>>>>>>> REPLACE/s;
+    let codeIsSearchReplaceBlock = searchReplaceRegex.test(code);
+
+    if (!codeIsSearchReplaceBlock) {
+      return; // only show apply button for search/replace code blocks
+    }
+
+    this.codeActions.push(
+      new TrackedObject({
+        fileUrl,
+        code,
+        containerElement: codeActionContainerElement,
+        state: 'ready' as CodeAction['state'],
+      }),
+    );
+  };
+
+  private patchCodeTask = task(async (codeAction: CodeAction) => {
+    codeAction.state = 'applying';
+    try {
+      let source = await this.cardService.getSource(
+        new URL(codeAction.fileUrl),
+      );
+
+      let applySearchReplaceBlockCommand = new ApplySearchReplaceBlockCommand(
+        this.commandService.commandContext,
+      );
+
+      let { resultContent: patchedCode } =
+        await applySearchReplaceBlockCommand.execute({
+          fileContent: source,
+          codeBlock: codeAction.code,
+        });
+
+      await this.cardService.saveSource(
+        new URL(codeAction.fileUrl),
+        patchedCode,
+      );
+      this.loaderService.reset();
+
+      codeAction.state = 'applied';
+    } catch (error) {
+      console.error(error);
```

Comment:
It's currently not clear what kind of errors we can anticipate here, so until we have a better idea we'll just log it to the console.